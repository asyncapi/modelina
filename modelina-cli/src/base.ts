import { Command } from '@oclif/core';

export default abstract class extends Command {
  async catch(err: Error & { exitCode?: number; }): Promise<any> {
    try {
      return await super.catch(err);
    } catch (e: any) {
      if (e instanceof Error) {
        this.logToStderr(`${e.name}: ${e.message}`);
        process.exitCode = 1;
      }
    }
  }
}
