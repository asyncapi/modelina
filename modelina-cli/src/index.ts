export {run} from '@oclif/core'

export {generateModels, ModelinaArgs, ModelinaFlags, Languages} from './helpers/generate'