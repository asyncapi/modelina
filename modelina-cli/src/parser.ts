import { AvroSchemaParser } from '@asyncapi/avro-schema-parser';
import { OpenAPISchemaParser } from '@asyncapi/openapi-schema-parser';
import { Parser, convertToOldAPI } from '@asyncapi/parser/cjs';
import { RamlDTSchemaParser } from '@asyncapi/raml-dt-schema-parser';
import { Flags } from '@oclif/core';
import { ProtoBuffSchemaParser } from '@asyncapi/protobuf-schema-parser';
import { getDiagnosticSeverity } from '@stoplight/spectral-core';
import { OutputFormat } from '@stoplight/spectral-cli/dist/services/config';
import { html, json, junit, pretty, stylish, teamcity, text } from '@stoplight/spectral-formatters';

import type { Diagnostic } from '@asyncapi/parser/cjs';
import type Command from './base';
import type { Specification } from './models/SpecificationFile';

export type SeveritytKind = 'error' | 'warn' | 'info' | 'hint';

export { convertToOldAPI };

const parser = new Parser({
  __unstable: {
    resolver: {
      cache: false,
    }
  }
});

parser.registerSchemaParser(AvroSchemaParser());
parser.registerSchemaParser(OpenAPISchemaParser());
parser.registerSchemaParser(RamlDTSchemaParser());
parser.registerSchemaParser(ProtoBuffSchemaParser());

export interface ValidationFlagsOptions {
  logDiagnostics?: boolean;
}

export function validationFlags({ logDiagnostics = true }: ValidationFlagsOptions = {}) {
  return {
    'log-diagnostics': Flags.boolean({
      description: 'log validation diagnostics or not',
      default: logDiagnostics,
      allowNo: true,
    }),
    'diagnostics-format': Flags.enum({
      description: 'format to use for validation diagnostics',
      options: Object.values(OutputFormat),
      default: OutputFormat.STYLISH,
    }),
    'fail-severity': Flags.enum<SeveritytKind>({
      description: 'diagnostics of this level or above will trigger a failure exit code',
      options: ['error', 'warn', 'info', 'hint'],
      default: 'error',
    }),
  };
}

interface ValidateOptions {
  'log-diagnostics'?: boolean;
  'diagnostics-format'?: `${OutputFormat}`;
  'fail-severity'?: SeveritytKind;
}

export async function validate(command: Command, specFile: Specification, options: ValidateOptions = {}) {
  const diagnostics = await parser.validate(specFile.text(), { source: specFile.getSource() });
  return logDiagnostics(diagnostics, command, specFile, options);
}

export async function parse(command: Command, specFile: Specification, options: ValidateOptions = {}) {
  const { document, diagnostics } = await parser.parse(specFile.text(), { source: specFile.getSource() });
  const status = logDiagnostics(diagnostics, command, specFile, options);
  return { document, diagnostics, status };
}

function logDiagnostics(diagnostics: Diagnostic[], command: Command, specFile: Specification, options: ValidateOptions = {}): 'valid' | 'invalid' {
  const logDiagnostics = options['log-diagnostics'];
  const failSeverity = options['fail-severity'] ?? 'error';
  const diagnosticsFormat = options['diagnostics-format'] ?? 'stylish';

  const sourceString = specFile.toSourceString();
  if (diagnostics.length) {
    if (hasFailSeverity(diagnostics, failSeverity)) {
      if (logDiagnostics) {
        command.logToStderr(`\n${sourceString} and/or referenced documents have governance issues.`);
        command.logToStderr(formatOutput(diagnostics, diagnosticsFormat, failSeverity));
      }
      return 'invalid';
    }

    if (logDiagnostics) {
      command.log(`\n${sourceString} is valid but has (itself and/or referenced documents) governance issues.`);
      command.log(formatOutput(diagnostics, diagnosticsFormat, failSeverity));
    }
  } else if (logDiagnostics) {
    command.log(`\n${sourceString} is valid! ${sourceString} and referenced documents don't have governance issues.`);
  }

  return 'valid';
}

export function formatOutput(diagnostics: Diagnostic[], format: `${OutputFormat}`, failSeverity: SeveritytKind) {
  const options = { failSeverity: getDiagnosticSeverity(failSeverity) };
  switch (format) {
  case 'stylish': return stylish(diagnostics, options);
  case 'json': return json(diagnostics, options);
  case 'junit': return junit(diagnostics, options);
  case 'html': return html(diagnostics, options);
  case 'text': return text(diagnostics, options);
  case 'teamcity': return teamcity(diagnostics, options);
  case 'pretty': return pretty(diagnostics, options);
  default: return stylish(diagnostics, options);
  }
}

function hasFailSeverity(diagnostics: Diagnostic[], failSeverity: SeveritytKind) {
  const diagnosticSeverity = getDiagnosticSeverity(failSeverity);
  return diagnostics.some(diagnostic => diagnostic.severity <= diagnosticSeverity);
}
