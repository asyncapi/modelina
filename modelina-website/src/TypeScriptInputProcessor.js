exports.TypeScriptInputProcessor = void 0;
const AbstractInputProcessor_1 = require("@asyncapi/modelina");
class TypeScriptInputProcessor extends AbstractInputProcessor_1.AbstractInputProcessor {
    shouldProcess(input) {
        return false;
    }
    process(input, options) {
        return Promise.reject();
    }
}
exports.TypeScriptInputProcessor = TypeScriptInputProcessor;
TypeScriptInputProcessor.settings = {
    uniqueNames: false,
    required: true,
    compilerOptions: {
        strictNullChecks: false
    }
};
//# sourceMappingURL=TypeScriptInputProcessor.js.map