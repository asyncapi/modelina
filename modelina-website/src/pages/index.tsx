
import Button from '@/components/buttons/Button';
import GithubButton from '@/components/buttons/GithubButton';
import CodeBlock from '@/components/CodeBlock';
import IconRocket from '@/components/icons/Rocket';
import GenericLayout from '@/components/layouts/GenericLayout';
import Heading from '@/components/typography/Heading';
import Paragraph from '@/components/typography/Paragraph';
import Image from 'next/image';

export default function Index() {
  const description =
    'Sometimes you just want to generate data models for your payload. Modelina is a library for generating data models based on inputs such as AsyncAPI, OpenAPI, or JSON Schema documents.';
  const image = '/img/social/modelina-card.jpg';

  return (
    <GenericLayout
      title="Modelina"
      description={description}
      image={image}
      wide
    >
      <div className="py-16 overflow-hidden lg:py-24">
        <div className="relative text-center">
          <Heading level="h1" typeStyle="heading-lg">
            Modelina
          </Heading>
          <Paragraph className="mt-4 max-w-3xl mx-auto">
            {description}
          </Paragraph>
        </div>

        <div className="relative mt-12 lg:mt-20 lg:grid lg:grid-cols-2 lg:gap-8 lg:items-center">
          <div className="relative mb-8 lg:mt-8">
            <Heading level="h4" typeStyle="heading-md-semibold">
              AsyncAPI Cli
            </Heading>
            <Paragraph className="mt-3 lg:pr-4">
              Get started immediately with Modelina through the AsyncAPI Cli.
            </Paragraph>
            <div className="mt-8">
              <CodeBlock
                language="generator-cli"
                showLineNumbers={false}
                className="mt-8"
                textSizeClassName="text-sm"
              >
                asyncapi generate models typescript ./asyncapi.json
              </CodeBlock>
              <div className="mt-8">
                <GithubButton
                  className="block mt-2 md:mt-0 md:inline-block w-full sm:w-auto mt-8"
                  href="https://github.com/asyncapi/cli"
                />
              </div>
            </div>
          </div>
          <div className="relative lg:mt-8 h-full">
            <Image
              src={'/img/card/cli-card.jpg'}
              fill
              sizes='100vw'
              alt={'CLI card'}
            />
          </div>
        </div>

        <div className="relative mt-12 lg:mt-20 lg:grid lg:grid-cols-2 lg:gap-8 lg:items-center">
          <div className="relative mb-8 lg:mt-8">
            <Heading level="h4" typeStyle="heading-md-semibold">
              Installation
            </Heading>
            <Paragraph className="mt-3 lg:pr-4">
              Install Modelina as a library to take full control.
            </Paragraph>
            <div className="mt-8">
              <CodeBlock
                language="bash"
                showLineNumbers={false}
                className="mt-8"
                textSizeClassName="text-sm"
              >
                npm install @asyncapi/modelina
              </CodeBlock>
              <div className="mt-8">
                <Button
                  className="hidden mt-2 md:mt-0 lg:inline-block md:ml-2"
                  text="Try The Playground"
                  href="/playground"
                  icon={<IconRocket className="inline-block -mt-1 w-6 h-6" />}
                />
              </div>
            </div>
          </div>
          <div className = "relative lg:mt-8 h-full">
            <Image 
            src = {'/img/card/modelina-card.jpg'}
            fill 
            sizes='100vw'
            alt = {'Modelina card'}
          />
          </div>
        </div>
        
        <div className="relative text-center mt-12">
          <Heading level="h1" typeStyle="heading-lg">
            Usage
          </Heading>
          <Paragraph className="mt-4 max-w-3xl mx-auto">
            Modelina can be used in a wide range of scenarios, and is build to
            be apart of the core toolbox you can customize to your hearts
            desire.{' '}
            <b>Here is just a fraction of what you can do with Modelina.</b>
          </Paragraph>
        </div>

        <div className="relative lg:mt-20 lg:items-center">
          <table className="min-w-full relative lg:mt-8">
            <tbody>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Modelina lets you generate data models from many types of{' '}
                    <a href="#inputs">inputs</a>
                  </b>
                </td>
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const asyncapi = ...
const jsonschema = ...
const openapi = ...
const metamodel = ...
...
const models = await generator.generate(
  asyncapi | jsonschema | openapi | metamodel
);`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Use the same inputs across a range of different{' '}
                    <a href="#outputs">generators</a>
                  </b>
                </td>
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const generator = new TypeScriptGenerator();
const generator = new CsharpGenerator();
const generator = new JavaGenerator();
const generator = new RustGenerator();
...
const models = await generator.generate(input);`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>Easily let you interact with the generated models.</b>-
                  Want to show the generated models on a website? Sure! - Want
                  to generate the models into files? Sure! - Want to combine all
                  the models into one single file? Sure! Whatever interaction
                  you need, you can create.
                </td>
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const models = await generator.generate(input);
for (const model in models) { 
  const generatedCode = generatedModel.result;
  const dependencies = generatedModel.dependencies;
  const modeltype = generatedModel.model.type;
  const modelName = generatedModel.modelName;
  ...
}`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Easily modify how models are{' '}
                    <a href="./docs/constraints.md">constrained</a> into the
                    output
                  </b>
                </td>

                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const generator = new TypeScriptGenerator({
  constraints: {
    modelName: ({modelName}) => {
      // Implement your own constraining logic
      return modelName;
    }
  }
});`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Seamlessly layer additional or replacement code{' '}
                    <a href="./docs/presets.md">
                      on top of each other to customize the models
                    </a>{' '}
                    to your use-case
                  </b>
                </td>

                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const generator = new TypeScriptGenerator({
  presets: [
    {
      class: {
        additionalContent({ content }) {
          return \`\${content}
public myCustomFunction(): string {
  return 'A custom function for each class';
}\`;
        },
      }
    }
  ]
});
const models = await generator.generate(input);`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Seamlessly lets you{' '}
                    <a href="./docs/presets.md">
                      combine multiple layers of additional or replacement code
                    </a>
                  </b>
                </td>

                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const myCustomFunction1 = {
  class: {
    additionalContent({ content }) {
      return \`\${content}
public myCustomFunction(): string {
return 'A custom function for each class';
}\`;
    },
  }
};
const myCustomFunction2 = {...};
const generator = new TypeScriptGenerator({
  presets: [
    myCustomFunction1,
    myCustomFunction2
  ]
});
const models = await generator.generate(input);`}
                  </CodeBlock>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </GenericLayout>
  );
}
