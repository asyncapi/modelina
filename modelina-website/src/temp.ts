import { InputMetaModel, ProcessorOptions, AbstractInputProcessor } from '@asyncapi/modelina';

/** Class for processing Typescript code inputs to Common module*/
export interface TypeScriptInputProcessorOptions {
  uniqueNames?: boolean;
  required?: boolean;
}
export class TypeScriptInputProcessor extends AbstractInputProcessor {
  shouldProcess(input: any): boolean {
    return false;
  }

  process(input: any, options?: ProcessorOptions): Promise<InputMetaModel> {
    return Promise.reject();
  }
}
