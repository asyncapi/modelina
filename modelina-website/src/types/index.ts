import { ParsedUrlQuery } from 'querystring';

export interface ModelProps {
  code: string;
  name: string;
}

export interface UpdateMessage {
  models: ModelProps[];
}

export interface ModelinaTypeScriptOptions {
  tsMarshalling: boolean;
  tsModelType: 'class' | 'interface' | undefined;
  tsEnumType: 'union' | 'enum' | undefined;
  tsModuleSystem: 'ESM' | 'CJS' | undefined;
  tsIncludeDescriptions: boolean;
}
export interface ModelinaJavaOptions {}
export interface ModelinaCplusplusOptions {}
export interface ModelinaGoOptions {}
export interface ModelinaJavaScriptOptions {}
export interface ModelinaCSharpOptions {
  csharpArrayType: 'List' | 'Array' | undefined;
  csharpAutoImplemented: boolean;
}
export interface ModelinaKotlinOptions {}
export interface ModelinaRustOptions {}
export interface ModelinaPythonOptions {}
export interface ModelinaDartOptions {}
export interface ModelinaGeneralOptions {
  language:
    | 'typescript'
    | 'java'
    | 'go'
    | 'javascript'
    | 'csharp'
    | 'kotlin'
    | 'rust'
    | 'python'
    | 'dart'
    | 'cplusplus';
}
export interface ModelinaGeneralQueryOptions {
  language: string;
}

export interface ModelinaJavaQueryOptions {}
export interface ModelinaGoQueryOptions {}
export interface ModelinaJavaScriptQueryOptions {}
export interface ModelinaCSharpQueryOptions {
  csharpArrayType?: string;
  csharpAutoImplemented?: string;
}
export interface ModelinaKotlinQueryOptions {}
export interface ModelinaRustQueryOptions {}
export interface ModelinaPythonQueryOptions {}
export interface ModelinaCplusplusQueryOptions {}
export interface ModelinaDartQueryOptions {}

export interface ModelinaTypeScriptQueryOptions {
  tsMarshalling?: string;
  tsModelType?: string;
  tsEnumType?: string;
  tsIncludeDescriptions?: string;
}

export interface ModelinaOptions
  extends ModelinaGeneralOptions,
    ModelinaTypeScriptOptions,
    ModelinaJavaOptions,
    ModelinaGoOptions,
    ModelinaJavaScriptOptions,
    ModelinaCSharpOptions,
    ModelinaKotlinOptions,
    ModelinaRustOptions,
    ModelinaPythonOptions,
    ModelinaDartOptions {}
export interface ModelinaQueryOptions
  extends ParsedUrlQuery,
    ModelinaGeneralQueryOptions,
    ModelinaTypeScriptQueryOptions,
    ModelinaJavaQueryOptions,
    ModelinaGoQueryOptions,
    ModelinaJavaScriptQueryOptions,
    ModelinaCSharpQueryOptions,
    ModelinaKotlinQueryOptions,
    ModelinaRustQueryOptions,
    ModelinaPythonQueryOptions,
    ModelinaCplusplusQueryOptions,
    ModelinaDartQueryOptions {}

export interface GenerateMessage extends ModelinaOptions {
  input: string;
}
export const modelinaLanguageOptions = [
  {
    value: 'typescript',
    text: 'TypeScript'
  },
  {
    value: 'java',
    text: 'Java'
  },
  {
    value: 'go',
    text: 'Go'
  },
  {
    value: 'javascript',
    text: 'JavaScript'
  },
  {
    value: 'csharp',
    text: 'C#'
  },
  {
    value: 'kotlin',
    text: 'Kotlin'
  },
  {
    value: 'rust',
    text: 'Rust'
  },
  {
    value: 'python',
    text: 'Python'
  },
  {
    value: 'dart',
    text: 'Dart'
  },
  {
    value: 'cplusplus',
    text: 'C++'
  }
];

export const defaultAsyncapiDocument = {
  asyncapi: '2.5.0',
  info: {
    title: 'Streetlights API',
    version: '1.0.0',
    description:
      'The Smartylighting Streetlights API allows you\nto remotely manage the city lights.\n',
    license: {
      name: 'Apache 2.0',
      url: 'https://www.apache.org/licenses/LICENSE-2.0'
    }
  },
  servers: {
    mosquitto: {
      url: 'mqtt://test.mosquitto.org',
      protocol: 'mqtt'
    }
  },
  channels: {
    'light/measured': {
      publish: {
        summary:
          'Inform about environmental lighting conditions for a particular streetlight.',
        operationId: 'onLightMeasured',
        message: {
          name: 'LightMeasured',
          payload: {
            type: 'object',
            $id: 'LightMeasured',
            properties: {
              id: {
                type: 'integer',
                minimum: 0,
                description: 'Id of the streetlight.'
              },
              lumens: {
                type: 'integer',
                minimum: 0,
                description: 'Light intensity measured in lumens.'
              },
              sentAt: {
                type: 'string',
                format: 'date-time',
                description: 'Date and time when the message was sent.'
              }
            }
          }
        }
      }
    },
    'turn/on': {
      subscribe: {
        summary:
          'Command a particular streetlight to turn the lights on or off.',
        operationId: 'turnOn',
        message: {
          name: 'TurnOn',
          payload: {
            type: 'object',
            $id: 'TurnOn',
            properties: {
              id: {
                type: 'integer',
                minimum: 0,
                description: 'Id of the streetlight.'
              },
              sentAt: {
                type: 'string',
                format: 'date-time',
                description: 'Date and time when the message was sent.'
              }
            }
          }
        }
      }
    }
  }
};
