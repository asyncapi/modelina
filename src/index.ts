export * from './generators';
export * from './helpers';
export * from './models';
export * from './processors';
export type { ModelLoggingInterface } from './utils';
export { Logger } from './utils';
