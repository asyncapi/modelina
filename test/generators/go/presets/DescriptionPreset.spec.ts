import { GO_DESCRIPTION_PRESET, GoGenerator } from '../../../../src';

const doc = {
  $id: 'Test',
  type: 'object',
  additionalProperties: false,
  required: ['string prop'],
  description: 'Main Description',
  properties: {
    'string prop': { type: 'string' },
    numberProp: {
      type: 'number',
      description: 'Description'
    },
    objectProp: {
      type: 'object',
      additionalProperties: false,
      $id: 'NestedTest',
      properties: { stringProp: { type: 'string', description: 'string prop' } }
    }
  }
};

describe('Description generation', () => {
  let generator: GoGenerator;
  beforeEach(() => {
    generator = new GoGenerator({
      presets: [GO_DESCRIPTION_PRESET]
    });
  });

  test('should render example function for model', async () => {
    const models = await generator.generate(doc);
    expect(models).toHaveLength(2);
    expect(models[0].result).toMatchSnapshot();
    expect(models[1].result).toMatchSnapshot();
  });
});
