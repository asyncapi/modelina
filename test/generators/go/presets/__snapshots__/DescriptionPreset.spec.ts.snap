// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description generation should render example function for model 1`] = `
"// Main Description
type Test struct {
  StringProp string
  // Description
  NumberProp float64
  ObjectProp *NestedTest
}"
`;

exports[`Description generation should render example function for model 2`] = `
"type NestedTest struct {
  // string prop
  StringProp string
}"
`;
